src/test/regress/expected/foreign_key_restriction_enforcement.out

```diff
@@ -0,0 +1,1089 @@
+DEBUG:  building index "on_update_fkey_table_pkey" on table "on_update_fkey_table"
+DEBUG:  validating foreign key constraint "fkey"
+ROLLBACK;
+-- case 3.7: DDL to a reference table is followed by COPY
+BEGIN;
+	ALTER TABLE referece_table  ADD COLUMN X int;
+DEBUG:  switching to sequential query execution mode
+DETAIL:  Reference relation "referece_table" is modified, which might have cascaded to hash distributed relations via foreign keys. Any parallel modification to those hash distributed relations in the same transaction can only be executed in sequential query execution mode
+	COPY on_update_fkey_table FROM STDIN WITH CSV;
+ERROR:  cannot execute parallel COPY on relation "on_update_fkey_table" since it holds conflicting locks via foreign keys on reference relation "referece_table" due to a previous "DDL" access in the same transaction 
+CONTEXT:  COPY on_update_fkey_table, line 2: "1002,99"
+ROLLBACK;
+-- case 3.8: DDL to a reference table is followed by TRUNCATE
+BEGIN;
+	ALTER TABLE referece_table  ADD COLUMN X int;
+DEBUG:  switching to sequential query execution mode
+DETAIL:  Reference relation "referece_table" is modified, which might have cascaded to hash distributed relations via foreign keys. Any parallel modification to those hash distributed relations in the same transaction can only be executed in sequential query execution mode
+	TRUNCATE on_update_fkey_table;
+DEBUG:  building index "on_update_fkey_table_pkey" on table "on_update_fkey_table"
+ROLLBACK;
+-- case 3.9: DDL to a reference table is followed by TRUNCATE
+BEGIN;
+	ALTER TABLE referece_table ALTER COLUMN id SET DATA TYPE smallint;
+DEBUG:  rewriting table "referece_table"
+DEBUG:  building index "referece_table_pkey" on table "referece_table"
+DEBUG:  validating foreign key constraint "fkey"
+	TRUNCATE on_update_fkey_table;
+DEBUG:  building index "on_update_fkey_table_pkey" on table "on_update_fkey_table"
+ROLLBACK;
+-----
+--- Now, start testing the other way araound
+-----
+-- case 4.1: SELECT to a dist table is follwed by a SELECT to a reference table
+BEGIN;
+	SELECT count(*) FROM on_update_fkey_table WHERE value_1 = 99;
+ count 
+-------
+    10
+(1 row)
+
+	SELECT count(*) FROM referece_table;
+ count 
+-------
+   101
+(1 row)
+
+ROLLBACK;
+-- case 4.2: SELECT to a dist table is follwed by a DML to a reference table
+BEGIN;
+	SELECT count(*) FROM on_update_fkey_table WHERE value_1 = 99;
+ count 
+-------
+    10
+(1 row)
+
+	UPDATE referece_table SET id = 101 WHERE id = 99;
+DEBUG:  switching to sequential query execution mode
+DETAIL:  Reference relation "referece_table" is modified, which might have cascaded to hash distributed relations via foreign keys. Any parallel modification to those hash distributed relations in the same transaction can only be executed in sequential query execution mode
+ERROR:  attempt to set query execution mode to sequential failed since a parallel query has already been executed in the same transaction
```

Comment:
it's a pity that we have to give up here

anyway, the message seems a hard to digest, especially without the debug information.

Maybe:
```
ERROR: cannot modify reference table "referece_table" because there was a parallel operation on a distributed table with a foreign key to "referece_table" in the same transaction
HINT: ...
```

Reply:
Yes, but the parallel access doesn't need to be on a relation which has a foreign key to the reference table. The following is a comment from the code:
```
/*
* If there has already been a parallel query executed, the sequential mode
* would still use the already opened parallel connections to the workers,
* thus contradicting our purpose of using sequential mode.
*/
```

So, it could be a completely unrelated parallel access to a colocated table. Are you OK with this restriction? That is probably the most restrictive part of the PR, but, the fix seemed to be touching the connection management, so I've not gone into that for now.

Reply:
It's not nice, but I wouldn't spend any time trying to fix it in 7.5.

I think we should do brainstorming on how to handle foreign keys to reference tables going forward.

Reply:
Yes, I agree with that. We've ended up with many restrictions. Let's first see which piece of these restrictions will bite users more..